discover/service.go

```diff
@@ -0,0 +1,655 @@
+// * receiving announcements validate them and apply the changes to router.
+// * sync lightning network with newly connected nodes with.
+// * broadcast announcements after validation.
+// * negotiate the channel announcement proofs exchange.
+// * handle the premature announcements.
+type Discovery struct {
+	// Parameters which are needed to properly handle the start and stop
+	// of the service.
+	started uint32
+	stopped uint32
+	quit    chan bool
+	wg      sync.WaitGroup
+
+	// cfg is a copy of the configuration struct that the discovery service
+	// was initialized with.
+	cfg *Config
+
+	// newBlocks is a channel in which new blocks connected to the end of
+	// the main chain are sent over.
+	newBlocks <-chan *chainntnfs.BlockEpoch
+
+	// prematureAnnouncements maps a blockheight to a set of networkMsgs
+	// which are "premature" from our PoV. An message is premature if
+	// it claims to be anchored in a block which is beyond the current main
+	// chain tip as we know it. Premature network messages will be processed
+	// once the chain tip as we know it extends to/past the premature
+	// height.
+	//
+	// TODO(roasbeef): limit premature networkMsgs to N
+	prematureAnnouncements map[uint32][]*networkMsg
+
+	// networkMsgs is a channel that carries new network broadcasted
+	// message from outside the discovery service to be processed by the
+	// networkHandler.
+	networkMsgs chan *networkMsg
+
+	// syncRequests is a channel that carries requests to synchronize newly
+	// connected peers to the state of the lightning network topology from
+	// our PoV.
+	syncRequests chan *syncRequest
+
+	// bestHeight is the height of the block at the tip of the main chain
+	// as we know it.
+	bestHeight uint32
+
+	fakeSig *btcec.Signature
+}
+
+// ProcessRemoteAnnouncement sends a new remote announcement message along with
+// the peer that sent the routing message. The announcement will be processed then
+// added to a queue for batched tickled announcement to all connected peers.
+// Remote channel announcements should contain the announcement proof and be
+// fully validated.
+func (d *Discovery) ProcessRemoteAnnouncement(msg lnwire.Message,
+	src *btcec.PublicKey) error {
+
+	aMsg := &networkMsg{
+		msg:      msg,
+		isRemote: true,
+		peer:     src,
+	}
+
+	select {
+	case d.networkMsgs <- aMsg:
+		return nil
+	case <-d.quit:
+		return errors.New("discovery has been shutted down")
+	}
+}
+
+// ProcessLocalAnnouncement sends a new remote announcement message along with
+// the peer that sent the routing message. The announcement will be processed then
+// added to a queue for batched tickled announcement to all connected peers.
+// Local channel announcements not contain the announcement proof and should be
+// fully validated. The channels proofs will be included farther if nodes agreed
+// to announce this channel to the rest of the network.
+func (d *Discovery) ProcessLocalAnnouncement(msg lnwire.Message,
```

Comment:
As we discussed on the slack, current implementation not allow us to do this, because funding manager should send the announcement to the `discovery` right after the channel have been created in order to be able use it for constructing the payment patuch.

Instead we come up with the solution when funding manager will use the router directly, and discovery will be subscribe on router topology notifications with such approach discovery will validate only remote announcements, but I would like to make such shift in standalone PR.